### ISSUE TITLE
loss:nan from training starts

### ISSUE THREAD

**Issue by jenniferzhu** [pandeydivesh15/NER-using-Deep-Learning#2]
For the English data sets, loss is nan since training starts. I tried to match the Keras and tensorflow versions (except that I use CPU), but it doesn't help. I also tried different datasets or optimizer, none of them helps. Can you please share some thoughts on troubleshooting?

FYI, every output looks the same before m.train(epochs=10) in the "NER-using-Deep-Learning" notebook.

**Comment by pandeydivesh15** [comment]
I hope you have all spacy data downloaded (for getting vectors).

Different versions can surely be a problem.
But I also faced a similar problem while I was working on this.
Actually, value of ```dropout``` was creating problems.
Try replacing ```m.make_and_compile()``` by ```m.make_and_compile(units=100, dropout=0.0, regul_alpha=0.0001)``` in the main ipython notebook.
Also, for better results, if you have enough memory, you can increase word vector dim to 300. Change value of ```self.LEN_WORD_VECTORS```(process_data.py) to 300.

**Comment by jenniferzhu** [comment]
Thanks for the prompt reply, Divesh!
Do you mind checking which version of spacy library you are using? I did notice that my spacy gave slightly different results in the spacy tests, but no errors were reported.
And thank you again for the suggestions on Keras codes. I will try that and see what happens.

**Comment by pandeydivesh15** [comment]
You’re welcome @jenniferzhu 
I am using ```spacy==1.6.0``` currently.

**Comment by jenniferzhu** [comment]
Hi Divesh, I modified the codes as you suggested. It does solve the "loss: nan" issue, but it will predict all tokens to the category "O". I guess dropout helps with the imbalanced dataset. Is there some other ways you have done to balance the dateset?

**Comment by pandeydivesh15** [comment]
Are versions of keras and tensorflow matching with  ```Keras==1.2.1 and tensorflow==0.12.1```?

**Comment by pandeydivesh15** [comment]
You can try changing ```regul_alpha``` while keeping dropout 0. Also, if you have sufficient memory, try changing len of vectors to 300. It will boost up results.
Regarding imbalanced dataset, try to prune it using some conditions. For e.g. maximum length of sentence < threshold, atleast one entity in every sentence, etc.

**Comment by jenniferzhu** [comment]
These are great tips! I got similar results to you by reinstalling spacy and using your parameters! Thank you, Divesh! BTW, do you have any recommended webpages so that I can read and understand why and how to tune those parameters?

**Comment by pandeydivesh15** [comment]
Currently, I don't have any specific webpages(for NER) but if you like, you can go through these links.
[how_to_choose_a_neural_network's_hyper-parameters](http://neuralnetworksanddeeplearning.com/chap3.html#how_to_choose_a_neural_network's_hyper-parameters)
[distill.pub](https://distill.pub/)

**Comment by jenniferzhu** [comment]
This is great! Thanks Divesh. You’re an expert!

**Comment by pandeydivesh15** [comment]
definitely not an expert :smile:

**Comment by jenniferzhu** [comment]
@pandeydivesh15 You suggestions worked great for the English dataset, but the same issue occurred again when I switched datasets. I am trying to understand what your logics is when you tuned the model, in order to avoid the same prediction "O" for all cases. I tried to grid search with learning rates, but it did not help. Can you please share your logics to tune the model?

**Comment by pandeydivesh15** [comment]
Sorry for late replying.
I had no specific logic while tuning model. The arguments to  ```m.make_and_compile()``` played an important factor. The most important one being ```dropout```.
In my case, setting dropout helped me in Hindi dataset while it was failing for English.
